src/metrics.hpp

```diff
@@ -286,38 +293,89 @@ class Metrics : public Allocated {
 
     void get_snapshot(Snapshot* snapshot) const {
       ScopedMutex l(&mutex_);
-      hdr_histogram* h = histogram_;
-      for (size_t i = 0; i < thread_state_->max_threads(); ++i) {
-        histograms_[i].add(h);
+
+      // In the "no refresh" case (the default) fall back to the old behaviour; add per-thread
+      // timestamps to histogram_ (without any clearing of data) and return what's there.
+      if (refresh_interval_ == CASS_DEFAULT_HISTOGRAM_REFRESH_INTERVAL_NO_REFRESH) {
+
+        for (size_t i = 0; i < thread_state_->max_threads(); ++i) {
+          histograms_[i].add(histogram_);
+        }
+
+        if (histogram_->total_count == 0) {
+          // There is no data; default to 0 for the stats.
+          copy_snapshot(zero_snapshot_, snapshot);
+        } else {
+          build_and_copy_snapshot(histogram_, snapshot);
+        }
+        return;
       }
 
-      if (h->total_count == 0) {
-        // There is no data; default to 0 for the stats.
-        snapshot->max = 0;
-        snapshot->min = 0;
-        snapshot->mean = 0;
-        snapshot->stddev = 0;
-        snapshot->median = 0;
-        snapshot->percentile_75th = 0;
-        snapshot->percentile_95th = 0;
-        snapshot->percentile_98th = 0;
-        snapshot->percentile_99th = 0;
-        snapshot->percentile_999th = 0;
-      } else {
-        snapshot->max = hdr_max(h);
-        snapshot->min = hdr_min(h);
-        snapshot->mean = static_cast<int64_t>(hdr_mean(h));
-        snapshot->stddev = static_cast<int64_t>(hdr_stddev(h));
-        snapshot->median = hdr_value_at_percentile(h, 50.0);
-        snapshot->percentile_75th = hdr_value_at_percentile(h, 75.0);
-        snapshot->percentile_95th = hdr_value_at_percentile(h, 95.0);
-        snapshot->percentile_98th = hdr_value_at_percentile(h, 98.0);
-        snapshot->percentile_99th = hdr_value_at_percentile(h, 99.0);
-        snapshot->percentile_999th = hdr_value_at_percentile(h, 99.9);
+      // Refresh interval is in use.  If we've exceeded the interval clear histogram_,
+      // compute a new aggregate histogram and build (and cache) a new snapshot.  Otherwise
+      // just return the cached version.
+      uint64_t now = get_time_since_epoch_ms();
+      if (now - refresh_timestamp_ >= refresh_interval_) {
+
+        hdr_reset(histogram_);
```

Comment:
Something I found a bit funny is that if I expand `HistogramWithRefreshInterval` test to wait for the next snapshot but without any requests, i.e. adding to the end:
```
  test::Utils::msleep(1.2 * refresh_interval);
  histogram.get_snapshot(&snapshot);
  EXPECT_EQ(snapshot.min, 0);
  EXPECT_EQ(snapshot.max, 0);
  EXPECT_EQ(snapshot.median, 0);
  EXPECT_EQ(snapshot.percentile_75th, 0);
  EXPECT_EQ(snapshot.percentile_95th, 0);
  EXPECT_EQ(snapshot.percentile_98th, 0);
  EXPECT_EQ(snapshot.percentile_99th, 0);
  EXPECT_EQ(snapshot.percentile_999th, 0);
  EXPECT_EQ(snapshot.mean, 0);
  EXPECT_EQ(snapshot.stddev, 0);
```
This (unexpectedly?) fails because snapshot.min has value `INT64_MAX` - which comes from `hdr_reset`.  When this windowing change is not enabled the initial value of `min` is zero so I think having this value reported differently in a window when there are no requests will cause confusion for devs and possibly issues for downstream metricking systems. This seems like something that should be addressed before merging this change but happy to discuss.

Reply:
After looking at this I'm kinda wondering if we've been doing this wrong the whole time.

We're populating the snapshot with the results of hdr_min() which looks like this:

```
int64_t hdr_min(struct hdr_histogram* h)
{
    if (0 < hdr_count_at_index(h, 0))
    {
        return 0;
    }

    return non_zero_min(h);
}
```

Near as I can tell this function has no notion of "the histogram is empty", which causes it to always enter the "return non-zero" case.  The short-circuit return above covers the case in which we've actually recorded one or more values of zero in the histogram.  In that case we return zero, but in all other cases we return the initial value of INT64_MAX.  Upshot is a clear distinction between "we saw no values" and "we saw values and the smallest one was zero".

The snapshot struct munges these two cases together by returning zeroes in the empty case, but it seems far more disruptive to try and change that now vs. making the new refresh case work the same as the existing interface.  To that end we'll just return the zero Snapshot if we have an empty histogram.